erigon-lib/state/bps_tree.go

```diff
@@ -134,6 +168,57 @@ type Node struct {
 	di  uint64 // key ordinal number in kv
 }
 
+func encodeListNodes(nodes []Node, w io.Writer) error {
+	numBuf := make([]byte, 8)
+	binary.BigEndian.PutUint64(numBuf, uint64(len(nodes)))
+	if _, err := w.Write(numBuf); err != nil {
+		return err
+	}
+
+	for ni := 0; ni < len(nodes); ni++ {
+		if _, err := w.Write(nodes[ni].Encode()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func decodeListNodes(data []byte) ([]Node, error) {
+	count := binary.BigEndian.Uint64(data[:8])
+	nodes := make([]Node, count)
+	pos := 8
+	for ni := 0; ni < int(count); ni++ {
+		dp, err := (&nodes[ni]).Decode(data[pos:])
+		if err != nil {
+			return nil, fmt.Errorf("decode node %d: %w", ni, err)
+		}
+		pos += int(dp)
+	}
+	return nodes, nil
+}
+
+func (n Node) Encode() []byte {
+	buf := make([]byte, 8+2+len(n.key))
+	binary.BigEndian.PutUint64(buf[:8], n.di)
+	binary.BigEndian.PutUint16(buf[8:10], uint16(len(n.key)))
+	copy(buf[10:], n.key)
+	return buf
+}
+
+func (n *Node) Decode(buf []byte) (uint64, error) {
+	if len(buf) < 10 {
+		return 0, errors.New("short buffer (less than 10b)")
+	}
+	n.di = binary.BigEndian.Uint64(buf[:8])
+	l := int(binary.BigEndian.Uint16(buf[8:10]))
+	if len(buf) < 10+l {
+		return 0, errors.New("short buffer")
+	}
+	n.key = buf[10 : 10+l]
+	//madvise(k, len(k), MADV_WILL_NEED)
```

Comment:
let's keep `madv_random` - until we see `.bt` on `pprof`. also we can `madv_will_need` (or `normal`) in future for whole `.bt` file. Also we already have temporary `madv_normal` in `OpenBtreeIndexWithDecompressor`